### DISCUSSION TITLE
Charon-nm not able to find TLS certificate

### DISCUSSION THREAD

**Discussion by jonax69** [strongswan/strongswan#437] [Q&A] [answered]
Hi, maybe someone can help me trying to figure out with NetworkManager and Strongswan's IKEv2 EAP-TLS authentication. At work we have some firewall appliances acting as IKEv2 endpoints, and user authentication is done with Microsoft NPS (Radius). User certificates are enrolled by an internal CA and deployed either with autoenroll (Windows 10) or exported manually (Linux). 
There are no problems on Windows 10 clients and everyone connects and authenticates ok. The same with Linux clients when using ipsec service, but if we try to use the NetworkManager applet we run into trouble. 
It complains about not finding the "TLS peer certificate for" the identity of the user after validating the other vpn endpoint certificates and the radius' one. 
It's a pity that we must include the identity on the applet, but if we leave itself, the CN of the certificate will be used and our fw appliances won't interpret correctly the user, so no posterior radius check can be done and fails.

Here are some relevant files:

ipsec:
[ipsec.conf.txt](https://github.com/strongswan/strongswan/files/6615756/ipsec.conf.txt)
[ipsec.secrets.txt](https://github.com/strongswan/strongswan/files/6615760/ipsec.secrets.txt)
[charon_debug.log](https://github.com/strongswan/strongswan/files/6615823/charon_debug.log)

charon_nm:
[VPN_TUNEL.nmconnection.txt](https://github.com/strongswan/strongswan/files/6615940/VPN_TUNEL.nmconnection.txt)
[syslog.txt](https://github.com/strongswan/strongswan/files/6615864/syslog.txt)

Private key is decrypted and cert information is as attached:
[user_cert.openssl.txt](https://github.com/strongswan/strongswan/files/6622376/user_cert.openssl.txt)

I followed what was recommended [in this link of serverfault.com](https://serverfault.com/questions/968190/strongswan-eap-identity-no-trusted-certificate) changing one of the SANs with no luck.

Tests are done on a fresh Ubuntu 21.01 with Strongswan 5.9.1-1ubuntu1 and Strongswan-nm 5.9.1-1ubuntu1

Thanks in advance

EDIT: corrected error on pseudonimization in user_cert.openssl.txt as quoted by @tobiasbrunner

**Comment by tobiasbrunner** [comment, marked as answer]
Does the SAN actually match the identity you configured? (In your modified output it doesn't.) You could also try increasing the log level for _tls_ (via `charon-nm.syslog.daemon.tls` in strongswan.conf) to see what's going on during the handshake.

**Reply by jonax69** [reply]
I've edited the original post and corrected the error. Yes, they match. eap-identity is user.someone@endpoint.com and in the SAN we have email:user.someone@endpoint.com. 

I've tried increasing the level for tls, but I don't see any userful additional information. Some log lines:
[syslog_tls.txt](https://github.com/strongswan/strongswan/files/6622676/syslog_tls.txt)
I've also increased the level for ASN, and can confirm that the cert is loaded and that rfc822Name is user.someone@endpoint.com. 

The fact that ipsec service is able to connect with the same cert, private key and eap-identity surprises me, but I suppose that charon and charon-nm are completely independent processes.

**Reply by tobiasbrunner** [reply]
> The fact that ipsec service is able to connect with the same cert, private key and eap-identity surprises me, but I suppose that charon and charon-nm are completely independent processes.

You didn't mention that before. Yes, they are completely independent and differently configured. And I found what causes this issue. The certificate is currently only enumerated if the configured identity is the certificate's subject DN. I pushed a fix for that to the `437-nm-cert` branch.

**Reply by jonax69** [reply]
Wow! Thanks, Tobias. 
Sadly, I have started a fresh new ubuntu, cloned that branch, compiled it but I still get the same message `"no usable TLS client certificate found for 'user.someone@endpoint.com"`. 

Just to be sure, I have compiled `437-nm-cert` branch and also `NetworkManager-strongswan-1.5.2`, is it ok? 

> You didn't mention that before. 

I tried, that's the reason why I attached the ipsec stuff, but it's clear that I didn't manage to do so 😄

**Reply by tobiasbrunner** [reply]
> Just to be sure, I have compiled 437-nm-cert branch and also NetworkManager-strongswan-1.5.2, is it ok?

It's generally not a good idea to mix builds from sources with binary packages. In this case it might work because the NM plugin, which the latter installs, does not use any strongSwan libraries.  But it still pulls in other packages. So when building from sources, you have to make sure your binaries replace those installed by the packages (i.e. `charon-nm` and the libraries). So you have to build with at least `--prefix=/usr --libexecdir=/usr/lib` and `--enable-nm`. If you intend to use this on a production system, it's probably better to uninstall (or not install) all strongSwan packages and also build the NM plugin from sources (it's in `src/frontends/gnome`).

> I tried, that's the reason why I attached the ipsec stuff, but it's clear that I didn't manage to do so

Ah, sorry. For some reason I thought that was from the server side and didn't look at it.

**Reply by jonax69** [reply]
I compiled both `437-nm-cert` and `NetworkManager-strongswan-1.5.2`, I didn´t know if the latter was needed. I understand that I'd better uninstall it and try with src/frontends/gnome one. I've tried it now and still complains.
I've realised that the `437-nm-cert` message is slightly different from the previous one. The old one was 
`[TLS] no TLS peer certificate found for 'user.someone@endpoint.com', skipping client authentication`
and now is
`[TLS] no usable TLS client certificate found for 'user.someone@endpoint.com'`

I have compiled strongswan with Openssl/nm plugin as suggested in an old ubuntu thread. No strongswan/Networkmanager-strongswan binary packages used in this fresh ubuntu:
`./configure --sysconfdir=/etc --prefix=/usr --libexecdir=/usr/lib \
--disable-aes --disable-des --disable-md5 --disable-sha1 --disable-sha2 \
--disable-fips-prf --disable-gmp --enable-openssl --enable-nm --enable-agent \
--enable-eap-gtc --enable-eap-md5 --enable-eap-mschapv2 --enable-eap-identity --enable-eap-tls`

Edit: missing argument in configure command

**Reply by tobiasbrunner** [reply]
> I compiled both `437-nm-cert` and `NetworkManager-strongswan-1.5.2`

Ah, did you mean the source tarball from [here](https://download.strongswan.org/NetworkManager/)? If so, then that's the same thing as what you find in `src/frontends/gnome`.

> I've realised that the `437-nm-cert` message is slightly different from the previous one.

Yes, the TLS library changed quite a bit with [5.9.2](https://wiki.strongswan.org/projects/strongswan/wiki/592).

Unfortunately, I can't reproduce this. Works fine here with any SAN. Maybe it's an issue with your particular certificate or key.

**Reply by jonax69** [reply]
Don't worry, thanks for all your efforts! If I finally get it working I'll write back for future reference

**Reply by tobiasbrunner** [reply]
If you want, you can send me the certificate and key via email and I'll have a look (use [this key](http://keys.gnupg.net/pks/lookup?search=0x765FE26C6B467584&fingerprint=on&op=index) if you want to do so encrypted).

**Reply by jonax69** [reply]
success! it works as you stated. User error, I didn't change to the branch after cloning the repo 🤦 Sorry for making you lose your time.

Just to know, is there any chance that this feature can be added officially on future releases?

**Reply by tobiasbrunner** [reply]
No problem, great you got it working eventually.

> Just to know, is there any chance that this feature can be added officially on future releases?

Yes, it will be included in the next release (8dbf40d19a7d328406d8a063314d86e795214239 is now in master). However, it will take some time until that version makes it to distros.
